src/datasets/arrow_dataset.py

```diff
@@ -2863,32 +2822,29 @@ def apply_function_on_filtered_inputs(inputs, indices, check_same_num_examples=F
             if with_rank:
                 additional_args += (rank,)
             processed_inputs = function(*fn_args, *additional_args, **fn_kwargs)
+            processed_inputs = (
+                {k: v for k, v in processed_inputs.data.items() if k not in processed_inputs.keys_to_format}
+                if isinstance(processed_inputs, LazyDict)
+                else processed_inputs
+            )
             if update_data is None:
                 # Check if the function returns updated examples
                 update_data = isinstance(processed_inputs, (Mapping, pa.Table))
                 validate_function_output(processed_inputs, indices)
             if not update_data:
                 return None  # Nothing to update, let's move on
-            if self._format_type or input_columns:
-                inputs = self._getitem(
-                    key=(indices if isinstance(indices, int) else slice(indices[0], indices[-1] + 1)),
-                    format_type=None,
-                    format_columns=None,
-                    format_kwargs=None,
-                    decoded=False,
-                )
             if remove_columns is not None:
-                for column in remove_columns:
-                    # `function` can modify input in-place causing column to be already removed.
-                    if column in inputs:
-                        inputs.pop(column)
+                pa_inputs = pa_inputs.select(
+                    [i for i, column in enumerate(pa_inputs.column_names) if column not in remove_columns]
+                )
+            pa_inputs_dict = {k: v for k, v in zip(pa_inputs.column_names, pa_inputs.itercolumns())}
             if check_same_num_examples:
-                input_num_examples = len(inputs[next(iter(inputs.keys()))])
+                input_num_examples = len(pa_inputs)
                 processed_inputs_num_examples = len(processed_inputs[next(iter(processed_inputs.keys()))])
                 if input_num_examples != processed_inputs_num_examples:
                     raise NumExamplesMismatchError()
-            if isinstance(inputs, dict) and isinstance(processed_inputs, Mapping):
-                return {**inputs, **processed_inputs}
+            if isinstance(inputs, Mapping) and isinstance(processed_inputs, Mapping):
+                return {**pa_inputs_dict, **processed_inputs}
```

Comment:
```suggestion
                # The .map() transform *updates* the dataset:
                # the output dictionary contains both the the input data and the output data.
                # The output dictionary may contain Arrow values from `pa_inputs_dict` so that we can re-write them efficiently.
                return {**pa_inputs_dict, **processed_inputs}
```